### PR TITLE
Exclude normal (non-block) strings from commented-out-code detection

### LINKED ISSUE #2: Normal (non-block) strings can count as comments
The line `String allFilesRegex = "/*";` gets flagged as "commented out code".

### PR DESCRIPTION
**Pull request by ThanGerlek**
Resolves #2

### CODE REVIEW

**Comment by 19mdavenport** [comment]
Love the continuation of banana token theme. Did you have fun writing nonsense banana code? 

Looks like a good start. Were you planning on working on the code to pass this test case as well?

**Comment by ThanGerlek** [comment]
Yep and yep

**Review by frozenfrank — CHANGES_REQUESTED** [review]
I can see how this is intended to work, but I feel like the line-by-line approach in the code leads itself to bugs in the cases when multiple key actions occur on the same line.

I feel like this can be expressed more simply and more accurately by the following states:
* In a single line comment
* In a multi line comment
* In a block string
* In a regular string
* In code (everything else)

I'll try to make a state diagram to represent this along with the transitions that characterize it. Note that the key to the state machine is to process the files character by character (rather than line by line).

**Comment by ThanGerlek** [comment]
The line-by-line approach is because that's how we're defining code comments as opposed to normal comments. If the line ends in one of a fixed set of chars, like `;`, then we consider it as code

**Comment by frozenfrank** [comment]
> The line-by-line approach is because that's how we're defining code comments as opposed to normal comments. If the line ends in one of a fixed set of chars, like `;`, then we consider it as code

## Acknowledgement
@ThanGerlek I understand that is how it's currently being done, and I acknowledge that it's been good enough to get us this far and correctly score 99% of the cases (thank you @19mdavenport!).

## Response
However, I'm pointing out that this approach is not the one to correctly answer the remaining 1% (which are now apparently being surfaced as important). If those 1% of the corner cases are important, then we probably intend to correctly decide all of them. If those 1% of the cases are not important, then we shouldn't even be discussing the PR since the current code already covers the most common, anticipated kinds of repositories.

Therefore, the purpose of this PR needs to be to correctly decide **all** comment related decisions. In order to correctly decide **all** of the comment related corner cases, we need to do a character by character analysis that is capable of catching and properly handling **all** possible scenarios. A line by line approach simply doesn't align with the syntactic definition of Java comments and strings.

While we could make the algorithm more and more complex to handle more and more diverse corner cases, we will always be excluding and improperly covering some of them. The code will be simpler when we align it with the true definition of the syntax anyways.

## State Diagram
A state diagram can represent this cleanly. Unfortunately, I don't think mermaid is doing a very good job at visually representing the ideas on the screen, but the ideas are represented.

### Notes and Introduction
Remember the following concepts from CS 252:
* The input is only accepted if the machine ends in an "accepting" state.
    * In this diagram, that can only be reached by reaching EOF `\z` while in the "code," or either of the comment states
    * Reaching EOF from any other state means the strings/literals comments mean the code is incorrect and won't compile anyways.
* The transitions are labelled with regexes.
* This diagram models a _non-deterministic state machine_.
    * This could be easily converted into a deterministic state machine with more verbose transitions
    * Parsing techniques practices in CS 236 enable us to interpret these kinds of languages
* For our purposes, lines of code can be determined by counting the number of newlines `\n` encountered while in the "code" state.
    * This can be generalized to lines of comments too.
* We do not need to handle single quote characters since Java only allows those to represent individual `chars`.
* Java makes it easy to perform this kind of string/comment analysis since it only allows a small set of syntaxes (those listed here):
    * There is no support for string interpolation
    * Backticks are disallowed
* Regex syntax used:
    *  `(?!negative_assertion)` asserts that "negative_assertion" is _not_ the next item in the sequence
    * `[^klass]` represents any character _not_ in the `klass` character class.
    * `\n` represents the _newline_ character
    * `\z` represents _end of file_
* With a little bit of careful ordering of the transitions, it really is this easy to account for all the possible complexities relating to string and comment parsing.
* I'm really disappointed with the automatic visual layout of the items of the diagram. Feel free to recreate the diagram in a different tool that will make it more beautiful if you wish.

### State Diagram

```mermaid
stateDiagram-v2
direction LR

state code
state "Single-Line Comment" as sl
state "Block String" as bs
state "Multi-Line Comment" as ml
state "Regular String" as rs

[*] --> code

%% Single line comments
code --> sl: //
sl --> sl: [^\n]
sl --> code: \n

%% Mult-line comments
code --> ml: /*
ml --> ml: (?!*/)
ml --> code: */

%% Block string
code --> bs: """
bs --> bs: (?!""")
bs --> code: """

%% Regular string
code --> rs: "
rs --> rs: "" | [^"]
rs --> code: "
%% note right of rs
%%     Single quoted items are not necessary
%%     to handle because they can only ever 
%%     be a single char, according to Java.
%% end note

ml --> [*]: \z
code --> [*]: \z
sl --> [*]: \z

```

File: src/main/java/edu/byu/cs240/checkstyle/readability/CommentedCode.java [hunk 1]
```diff
@@ -17,6 +18,14 @@ public class CommentedCode extends AbstractFileSetCheck {
 
     private static final Set<Character> CODE_LINE_END_CHARS = Set.of(';', ',', '{', '}', '(', ')', '/');
 
+    // STRINGS_REGEX matches strings that begin with an even number of non-escaped " chars
+    private static final String STRINGS_REGEX = "^[^\"]*((?<!/)\"[^\"]*(?<!/)\"[^\"]*)*";
```
Comment (frozenfrank) [inline src/main/java/edu/byu/cs240/checkstyle/readability/CommentedCode.java:22]: Thanks for clearly explaining how this system is behaving!

File: src/test/java/edu/byu/cs240/checkstyle/readability/CommentedCodeTest.java [hunk 2]
```diff
@@ -53,6 +53,14 @@ public void should_FindNoErrors_when_TextBlock() throws CheckstyleException {
     }
 
 
+    @Test
+    @DisplayName("Should find no errors in code that has '/*' or related character sequences in a string")
+    public void should_FindNoErrors_when_commentInString() throws CheckstyleException {
+        String fileName = "testInputs/commentedCode/should_FindNoErrors_when_commentInString.java";
```
Comment (frozenfrank) [inline src/test/java/edu/byu/cs240/checkstyle/readability/CommentedCodeTest.java:59]: I love that you led with well crafted tests to verify that the code actually behaves as it should. Nice work!

File: src/test/resources/edu/byu/cs240/checkstyle/readability/testInputs/commentedCode/should_FindNoErrors_when_commentInString.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        String alsoNotABananaToken_withEndChar = "/**;";
+        String stillNotABananaToken_withEndChar = "//;";
+        String definitelyNotABananaToken_withEndChar = "///;";
+        String dontEvenThinkAboutThisBeingABananaToken_withEndChar = "////;";
+
+        String notABananaToken_twice_withEndChar = "/*;" + "/*;";
+        String alsoNotABananaToken_twice_withEndChar = "/**;" + "/**;";
+        String stillNotABananaToken_twice_withEndChar = "//;" + "//;";
+        String definitelyNotABananaToken_twice_withEndChar = "///;" + "///;";
+        String dontEvenThinkAboutThisBeingABananaToken_twice_withEndChar = "////;" + "////;";
```
Comment (frozenfrank) [inline src/test/resources/edu/byu/cs240/checkstyle/readability/testInputs/commentedCode/should_FindNoErrors_when_commentInString.java:27]: 🎶 🕺 "You _ want _ my _ banana? You _ can't _ have _ my _ banana!" — Sung to the tune of ["PB&J" from EFY/FSY](https://youtu.be/3YedczKa1f4?si=KVYCdQr8BLZSuh0s&t=3) 🕺

**Review by frozenfrank — APPROVED** [review]
Looks, great @ThanGerlek! Thanks for putting together these nice changes and proving they work with test cases as well 👍

**Comment by 19mdavenport** [comment]
Counter proposal, #5

**Comment by ThanGerlek** [comment]
Closing in favor of #5

**Pull request closed without merging**